### ISSUE TITLE
Upon reboot, AFWall does not start until phone is unlocked

### ISSUE THREAD

**Issue by markd89** [ukanth/afwall#1056]
I have a OP5 w/ LineageOS 16, MagiSK, no GAPPS.

AFWall v 3.3.1 Donate

Startup script /sbin/.core/img/.core/service.d

When my phone is rebooted, it boots up locked and I have to do a lock screen pattern to unlock it.
AFWall does not start and apply the rules until after the lock screen is unlocked.

Am I doing something wrong?

**Comment by Fotzenfigaro** [comment]
With Magisk try
/sbin/.magisk/img/.core/service.d

**Comment by markd89** [comment]
Thanks for the idea. 

I just tried that: Same result. No AFWall until after the phone is unlocked

**Comment by defarram** [comment]
On my phone the `/data/data/` directory is encrypted after boot until the screen is unlocked (using password).
The [afwallstart](https://github.com/ukanth/afwall/blob/93bc549c3fabd6abc7c0150a99c14e6ec135b959/app/src/main/res/raw/afwallstart#L5) script requires `/data/data/` directory to be available.
This is the reason why it does not work. But the fix is easy: Just use system iptables binary and ignore iptables binary from `/data/data/`.

On Android 7.0 and higher, the`/data/data/` directory is encrypted until entering password to unlock screen. This is called `Direct Boot mode`.

**Comment by MrEngineerMind** [comment]
OK, I am not an IPtables expert, so I have some questions...

I have a Pixel 4a Running Android 10.

1) In the "Startup Directory path for script" I am using "/data/adb/service.d". Since this is not "/data/data" does that mean your "doesnt work until unlocked" comment doesn't apply to me and my "startup leak" is working before being unlocked?

2) You then say to use the "system iptables and ignore iptables binary from /data/data". But when I go into the "binaries" setting in afwall, it only allows me to specify "system iptables" or "built-in iptables" - it doesn't allow me to specify a directory, so I am confused with the /data/data part of your statement.

Please help me understand what you have said.

**Comment by defarram** [comment]
>     1. In the "Startup Directory path for script" I am using "/data/adb/service.d". Since this is not "/data/data" does that mean your "doesnt work until unlocked" comment doesn't apply to me and my "startup leak" is working before being unlocked?
> 

No, it still applies because the [afwallstart](https://github.com/ukanth/afwall/blob/93bc549c3fabd6abc7c0150a99c14e6ec135b959/app/src/main/res/raw/afwallstart#L5) script which is automatically inserted into `/data/adb/service.d` directory has a bug. That script requires `/data/data/` directory to be available (unlocked/unencrypted).



>     2\. You then say to use the "system iptables and ignore iptables binary from /data/data". But when I go into the "binaries" setting in afwall, it only allows me to specify "system iptables" or "built-in iptables" - it doesn't allow me to specify a directory, so I am confused with the /data/data part of your statement.

I wrote about fixing the bug in that script, not specifying an options in settings. The script should not use iptables binary from `/data/data/` directory. The simplest fix is to remove line containing `exit` word from that script and after that it probably will work.

**Comment by ukanth** [comment]
the reason it was checking /data/data due to scenario when someone uninstall afwall with startup script installed. in that case they will not be able use internet due to DROP for all default chains.

It still uses /system/bin/iptables to set those.

**Comment by MrEngineerMind** [comment]
So, can this be done...

Modify the script so that it doesn't try to access /data/data so that the script can fully run even before the phone is unlocked.

And I understand that this script will be orphaned if afwall is uninstalled without disabling this leak feature, causing the device to not be able to access the internet.

But can Afwall be possibly modified so that if the user did uninstall afwall and locked out internet access, that they can simply reinstall afwall, disable the leak feature, then uninstall afwall again to restore internet access.

Can this be done?

If so, then I think this would be a good balance - have leak protection, but also have a way to recovery internet access if afwall was uninstalled without turning leak protection off first.

**Comment by ukanth** [comment]
The main reason it was introduced due to same problem people were complaining. If you want, you could always modify the startup script by yourself and remove the /data/data/ check. I will see If anything can be introduced without affecting this check.

**Comment by defarram** [comment]
Now, I understand why this check was introduced.

On my phone these files are available before entering password and unlocking the screen:
```
./data_mirror/cur_profiles/0/dev.ukanth.ufirewall
./data_mirror/data_de/null/0/dev.ukanth.ufirewall
./config/sdcardfs/dev.ukanth.ufirewall
./data/misc/profiles/cur/0/dev.ukanth.ufirewall
./data/misc/profiles/ref/dev.ukanth.ufirewall
./data/app/~~NyEMuYPoxC2s8Jw1QwL-Hu==/dev.ukanth.ufirewall-DH-843SxMlWkD9WeLkJq2W==
./data/user_de/0/dev.ukanth.ufirewall
```

But I don't know which one of them we should check. Maybe not each of these files is available on every phone.

Here is explained what these random strings in `/data/app/` directory are: https://android.stackexchange.com/questions/215653/what-is-the-code-after-the-app-package-name-in-data-app

Probably the file in `/data/app/` is available on every phone, but it contains random string in the path so if we would like to use this file, we would have to generate afwallstart script containing correct path.

More details about Direct Boot from https://developer.samsungknox.com/blog/samsung-devices-and-file-based-encryption-fbe :
> FBE is a feature available with Android OS 7.0 and higher. FBE allows encryption of different files with different keys that are used to unlock each file independently.
> 
> An FBE-enabled device offers each user and all applications the following two storage options:
>
> Credential Encrypted (CE) storage: CE storage is the default storage location that is only available after the user unlocks the device.
> Device Encrypted (DE) storage: DE storage is a storage location available during both the Direct Boot mode as well as after the user unlocks the device.
> 
> FBE enables a new feature called Direct Boot that is available in Android OS 7.0 or higher. Direct Boot lets encrypted devices boot straight to the lock screen, resulting in apps operating within a limited context.
> 
> The introduction of FBE and related new APIs help applications become aware of encryption—making apps ‘crypto-aware’—and let these apps operate within a limited context. Examples of this situation include situations where the user restarts the device but has not yet provided their login credentials. In such a situation, the apps that are crypto-aware still work, but any private information is still under lockdown due to FBE. For more information on FBE and how it helps secure your device, see File-based Encryption.
